### PR TITLE
182409679 sim feature

### PR DESCRIPTION
**Pull request by androidmaven**
- Sim page with compose.
- New codes use the data-domain-presentation architecture
- Auto update channels db from json file if required after app updates.

N.B
This PR does not yet include: showing actual sim first before displaying it's account information. It currently functions as it is previously, however with code updates to use the new architecture. Hat tip to @alexosugo's suggestion.

unsupported_sim_single

https://user-images.githubusercontent.com/21104847/187464000-213e2685-632f-4b2f-9e92-a29820e33a98.mp4

### CODE REVIEW

**Comment by davkutalek** [comment]
@androidmaven There shouldn't be any add flow for SIMs. Once we have the phone permission we automatically show all the users SIM cards.

File: app/src/main/java/com/hover/stax/accounts/AccountsViewModel.kt
Comment (alexosugo) [inline app/src/main/java/com/hover/stax/accounts/AccountsViewModel.kt:24]: All Sim functions should move to their own `ViewModel`. We're moving from function-based ViewModels to screen-based ViewModels. It's easier to track actions and side-effects that way. Use use-cases to compose your functionality as they are reusable.

File: app/src/main/java/com/hover/stax/addChannels/ChannelsViewModel.kt
Comment (alexosugo) [inline app/src/main/java/com/hover/stax/addChannels/ChannelsViewModel.kt:37]: Please refer to the above comment.

File: app/src/main/java/com/hover/stax/bonus/BonusViewModel.kt
Comment (alexosugo) [inline app/src/main/java/com/hover/stax/bonus/BonusViewModel.kt:21]: This class will eventually be removed in favor of the `GetBonusUseCase`. Make the changes there as well. Eventually, fetch bonus usages will be directed to the usecase.

File: app/src/main/java/com/hover/stax/data/repository/AccountRepositoryImpl.kt
Comment (alexosugo) [inline app/src/main/java/com/hover/stax/data/repository/AccountRepositoryImpl.kt:27]: You should use the repo not the usecase here. UseCases should be used only on the presentation layer. While testing, we provide mocks of the repos used.

File: app/src/main/java/com/hover/stax/data/repository/BonusRepositoryImpl.kt
Comment (alexosugo) [inline app/src/main/java/com/hover/stax/data/repository/BonusRepositoryImpl.kt:21]: Refer to above comment.

File: app/src/main/java/com/hover/stax/presentation/simcard/SimFragment.kt
Comment (alexosugo) [inline app/src/main/java/com/hover/stax/presentation/simcard/SimFragment.kt:31]: SimFragment should only use one viewmodel.

**Review by alexosugo — CHANGES_REQUESTED** [review]
Clicking continue on the dialog that shows up after clicking the link sim button does nothing. I'm not sure if this is a known bug.

File: app/src/main/java/com/hover/stax/accounts/AccountsViewModel.kt
Comment (androidmaven) [reply app/src/main/java/com/hover/stax/accounts/AccountsViewModel.kt:24]: Thanks for this. Code updated

File: app/src/main/java/com/hover/stax/addChannels/ChannelsViewModel.kt
Comment (androidmaven) [reply app/src/main/java/com/hover/stax/addChannels/ChannelsViewModel.kt:37]: Thanks!

File: app/src/main/java/com/hover/stax/bonus/BonusViewModel.kt
Comment (androidmaven) [reply app/src/main/java/com/hover/stax/bonus/BonusViewModel.kt:21]: Code updated.

File: app/src/main/java/com/hover/stax/data/repository/AccountRepositoryImpl.kt
Comment (androidmaven) [reply app/src/main/java/com/hover/stax/data/repository/AccountRepositoryImpl.kt]: Thanks for this.

File: app/src/main/java/com/hover/stax/presentation/simcard/SimFragment.kt
Comment (androidmaven) [reply app/src/main/java/com/hover/stax/presentation/simcard/SimFragment.kt]: Code updated.

File: app/src/main/java/com/hover/stax/domain/use_case/sims/GetLivePresentSimUseCase.kt
Comment (alexosugo) [inline app/src/main/java/com/hover/stax/domain/use_case/sims/GetLivePresentSimUseCase.kt:6]: This can be combined with the `GetPresentSimsUseCase` to form a single class, say `GetPresentSims` since they both do the same thing.

File: app/src/main/java/com/hover/stax/domain/use_case/sims/GetPresentSimUseCase.kt
Comment (alexosugo) [inline app/src/main/java/com/hover/stax/domain/use_case/sims/GetPresentSimUseCase.kt:7]: Please refer to the previous comment.

File: app/src/main/java/com/hover/stax/presentation/simcard/SimViewModel.kt
Comment (alexosugo) [inline app/src/main/java/com/hover/stax/presentation/simcard/SimViewModel.kt:32]: Use the `GetAccountsUseCase` to fetch these accounts. That way, you can abstract away the `AccountRepo`.

**Review by alexosugo — CHANGES_REQUESTED** [review]
After clicking the `Link Sim Cards` button, I get the confirmation dialog but nothing happens once I approve it.

**Comment by davkutalek** [comment]
@androidmaven my phone always shows the "Link your SIM cards to Stax" even tho I've granted the permissions

File: app/src/main/java/com/hover/stax/data/repository/AccountRepositoryImpl.kt
Comment (alexosugo) [inline app/src/main/java/com/hover/stax/data/repository/AccountRepositoryImpl.kt:39]: Things to note on this function:
- The account name allocation needs to be changed to match the one in `ChannelsViewModel`.
- Generally, batch inserts are faster than multiple single inserts. 
- Batch inserts are atomic, meaning if one insert/update fails, the entire transaction is rolled back, depending on use case, this may be a desired effect.

File: app/src/main/java/com/hover/stax/data/repository/ChannelRepositoryImpl.kt
Comment (alexosugo) [inline app/src/main/java/com/hover/stax/data/repository/ChannelRepositoryImpl.kt:15]: Duplicate method as the one in `SimRepositoryImpl`.

File: app/src/main/java/com/hover/stax/home/MainActivity.kt
Comment (alexosugo) [inline app/src/main/java/com/hover/stax/home/MainActivity.kt:87]: Consider using `StateFlow`. It makes doing this unnecessary.

File: app/src/main/java/com/hover/stax/presentation/simcard/SimScreen.kt
Comment (alexosugo) [inline app/src/main/java/com/hover/stax/presentation/simcard/SimScreen.kt:55]: There exists a usecase to fetch bonuses. Add it to `SimViewModel` and remove the `BonusViewModel`.

**Review by alexosugo — CHANGES_REQUESTED** [review]
It is functioning well now. A few things to consider:
- The `check balance` button has a blue background around it (other than the button color). Try increasing the brightness on your device to see it. 
- For all queries to the db, opt for `Flow` instead of `LiveData` as they observe database changes and propagate them as they occur. 
- For all viewmodels to be used in compose components, use `StateFlow` and `SharedFlow`. They remove the need to add empty observers and we remove the risk of the `multiple observers on same thread` bug.

**Review by alexosugo — APPROVED** [review]

**Comment by alexosugo** [comment]
@davkutalek Awaiting your final review. The feature should be functioning as intended now.

**Review by davkutalek — CHANGES_REQUESTED** [review]
Overall this is great. There are some basic refactoring to do in terms of de-duplicating the code for account creation. 

I had a couple other issues:
- When updating from my existing install it didn't add my accounts. It asked for the permissions (even though they were granted) and then didn't do anything.
- The permission ask is always at the top even when my SIM is detected and shown.
- All detected SIMs should be shown, but if they have no channel should just say "unsupported". This way it is clear that we have detected it.

File: app/src/main/java/com/hover/stax/presentation/sim/SimScreen.kt
Comment (davkutalek) [inline app/src/main/java/com/hover/stax/presentation/sim/SimScreen.kt:258]: It is strange that operatorName is empty, that is the correct variable. I have reproduced the issue on my device/SIM will look into how to fix it. This is fine in the meantime, just want some minor UI changes which I will describe.

**Review by davkutalek — APPROVED** [review]
@androidmaven this is great, just needs those minor UI changes as mentioned, see below. I will do some work on the SDK to fix the issue with the name, not a blocker for release.

UI changes:
- Change "Your linked SIM cards" to "Your SIM cards" and make it the title of the page (replace the "SIM" title above)
- For SIMs that aren't supported let's use the same layout with the telco name and slot number and icon at the top. I will find a way to make sure there is always a name, but if it is blank just put "Unknown" for now. The icon can be blank for now, but there is supposedly a way to get it from Android, I will work on this as well. In place of the check balance button just put the word unsupported. Then remove the first paragraph and the full text  can read "If this SIM supports USSD please reach out to our support team so we can integrate it." The email button should be the style of the buy airtime button (not blue).

**Comment by davkutalek** [comment]
@androidmaven looks good, only minor thing from me is that Unsupported should just be text not a button

**Pull request merged**
